ai-catalog/samples/gemini-live-todo/src/main/java/com/android/ai/samples/geminilivetodo/ui/TodoScreenViewModel.kt

```diff
@@ -0,0 +1,246 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.android.ai.samples.geminilivetodo.ui
+
+import android.Manifest
+import android.annotation.SuppressLint
+import android.app.Activity
+import android.content.pm.PackageManager
+import android.util.Log
+import androidx.core.app.ActivityCompat
+import androidx.core.content.ContextCompat
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.android.ai.samples.geminilivetodo.data.TodoRepository
+import com.google.firebase.Firebase
+import com.google.firebase.ai.ai
+import com.google.firebase.ai.type.FunctionCallPart
+import com.google.firebase.ai.type.FunctionDeclaration
+import com.google.firebase.ai.type.FunctionResponsePart
+import com.google.firebase.ai.type.GenerativeBackend
+import com.google.firebase.ai.type.LiveSession
+import com.google.firebase.ai.type.PublicPreviewAPI
+import com.google.firebase.ai.type.ResponseModality
+import com.google.firebase.ai.type.Schema
+import com.google.firebase.ai.type.SpeechConfig
+import com.google.firebase.ai.type.Tool
+import com.google.firebase.ai.type.Voice
+import com.google.firebase.ai.type.content
+import com.google.firebase.ai.type.liveGenerationConfig
+import dagger.hilt.android.lifecycle.HiltViewModel
+import javax.inject.Inject
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.SharingStarted
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.combine
+import kotlinx.coroutines.flow.stateIn
+import kotlinx.coroutines.launch
+import kotlinx.serialization.json.JsonObject
+import kotlinx.serialization.json.JsonPrimitive
+import kotlinx.serialization.json.jsonPrimitive
+import kotlinx.serialization.json.long
+
+@OptIn(PublicPreviewAPI::class)
+@HiltViewModel
+class TodoScreenViewModel @Inject constructor(private val todoRepository: TodoRepository) : ViewModel() {
+    private val TAG = "TodoScreenViewModel"
+    private var session: LiveSession? = null
+
+    private val liveSessionState = MutableStateFlow<LiveSessionState>(LiveSessionState.NotReady)
+    private val todos = todoRepository.todos
+
+    val uiState: StateFlow<TodoScreenUiState> = combine(liveSessionState, todos) { liveSessionState, todos ->
+        TodoScreenUiState.Success(todos, liveSessionState)
+    }.stateIn(
+        scope = viewModelScope,
+        started = SharingStarted.WhileSubscribed(5000L),
+        initialValue = TodoScreenUiState.Initial,
+    )
+
+    fun addTodo(taskDescription: String) {
+        todoRepository.addTodo(taskDescription)
+    }
+
+    fun removeTodo(todoId: Long) {
+        todoRepository.removeTodo(todoId)
+    }
+
+    fun toggleTodoStatus(todoId: Long) {
+        todoRepository.toggleTodoStatus(todoId)
+    }
+
+    @SuppressLint("MissingPermission")
+    fun toggleLiveSession(activity: Activity) {
```

Comment:
You shouldn't ever pass an Activity into the view model because it causes memory leaks.

Permissions requests are done in the UI layer